### ISSUE TITLE
[FR] Have seoImage populate twitterImage and ogImage tags when set via Twig

### ISSUE THREAD

**Issue by roberttolton** [nystudio107/craft-seomatic#253]
SEOMatic Version: `3.1.23`
Craft Version: `3.0.28`

Have just come across this issue where I am setting `seoImage` in the template, to whatever image the first of my image modules has:

```
{% set break = 0 %}
{% set seoImage = 'not-set' %}
{% for module in attribute(entry, modulesField).all() if break == 0 %}

    {% if module.type == 'imageModule' %}
        {% set break = 1 %}

        {% if entry.type.handle == 'post' and module.image | length %}
            {% do seomatic.meta.seoImage( module.image[0].url ) %}
            {% set seoImage = module.image[0].url %}
        {% else %}
            {% do seomatic.meta.seoImage( alias('@baseAssetUrl/imported-images/' ~ module.image) ) %}
            {% set seoImage = alias('@baseAssetUrl/imported-images/' ~ module.image) %}
        {% endif %}

        {% if module.imageDescription %}
            {% do seomatic.meta.seoDescription( seomatic.helper.truncateOnWord( module.imageDescription, 300 ) ) %}
        {% endif %}
    {% endif %}

{% endfor %}
```

The `seoImage` variable is purely for testing.

The output of my page is like so:

```
Set During Loop: {{ seoImage }}<br>
twitterImage (Set using: {% verbatim %}{% do seomatic.meta.twitterImage( module.image[0].url ) %}{% endverbatim %}): [{{ seomatic.meta.parsedValue('twitterImage') }}]<br>
ogImage (Set using: {% verbatim %}{% do seomatic.meta.ogImage( module.image[0].url ) %}{% endverbatim %}): [{{ seomatic.meta.parsedValue('ogImage') }}]<br>
seoImage (Set using: {% verbatim %}{% do seomatic.meta.seoImage( module.image[0].url ) %}{% endverbatim %}): [{{ seomatic.meta.parsedValue('seoImage') }}]
```

Which produces:

```
Set During Loop: http://craft-3.miraidev/uploads/image-module-images/2018/07/5324/DSC04892.jpg
twitterImage (Set using: {% do seomatic.meta.twitterImage( module.image[0].url ) %}): []
ogImage (Set using: {% do seomatic.meta.ogImage( module.image[0].url ) %}): []
seoImage (Set using: {% do seomatic.meta.seoImage( module.image[0].url ) %}): [http://craft-3.miraidev/uploads/image-module-images/2018/07/5324/DSC04892.jpg]
```

Note how the `twitterImage` and `ogImage` tags output nothing. This means that the right tags aren't inserted into `<head>`, even though `seoImage` does contain a value. If I set `twitterImage` and `ogImage` each manually, they will appear correctly.

The `twitter:card` `twitter:site` `twitter:creator` `twitter:title` `twitter:description` (and `og` equivalents) tags all appear, just not the `twitter:image` tag if I only try and set it with `do seomatic.meta.seoImage`

Within my SEOMatic settings, everything is set to fallback to the SEO Image.

--

Workaround currently is just to explicitly set `do seomatic.meta.ogImage` and `do seomatic.meta.twitterImage` in the template.

**Comment by khalwat** [comment]
So by default, the `ogImage` and `twitterImage` will just pull from whatever you've set your `seoImage` to... if that's your desired result, then just change the `seomatic.meta.seoImage` and away you go.

Now as for why it's not working for you by setting things explicitly; I'm a little confused about your use of the `{% verbatim %}` tag here -- anything inside of it, Twig will not process.

Can you clarify what's going on here by presenting the code where you're setting the `twitterImage`, `ogImage`, etc.?

**Comment by roberttolton** [comment]
The verbatim tags are just used for debugging, I just wanted to show what code was producing what result.

I’m setting the seoImage using the do function, but it just doesn’t trickle down to the others who are meant to inherit it. I’ve not set an seoImage within the plugin settings - I need to define it dynamic in the template.

**Comment by khalwat** [comment]
Right, but by default the `ogImage` and `twitterImage` will use whatever value you set the `seoImage` to -- so you shouldn't need to change that.

Can you show me the actual code where you're attempting to change the `ogImage` and the `twitterImage`?

**Comment by roberttolton** [comment]
@khalwat here's an abridged template, stripped of irrelevant pieces:

```
{% extends "_layout" %}

{% if entry.type.handle == 'post' %}
    {% set modulesField = 'modules' %}
{% else %}
    {% set modulesField = 'importedModules' %}
{% endif %}

{% set break = 0 %}
{% for module in attribute(entry, modulesField).all() if break == 0 %}

    {% if module.type == 'imageModule' %}
        {% set break = 1 %}

        {% if entry.type.handle == 'post' and module.image | length %}
            {% do seomatic.meta.seoImage( module.image[0].url ) %}
        {% else %}
            {% do seomatic.meta.seoImage( alias('@baseAssetUrl/imported-images/' ~ module.image) ) %}
        {% endif %}

        {% do seomatic.meta.twitterCard( 'summary_large_image' ) %}

        {% if module.imageDescription %}
            {% do seomatic.meta.seoDescription( seomatic.helper.truncateOnWord( module.imageDescription, 300 ) ) %}
        {% endif %}
    {% endif %}

{% endfor %}

{% block content %}

    {% include 'includes/_modules.twig' %}

{% endblock %}
```

You can see I'm setting `seoImage` using the `do` tag, however this doesn't propagate to the `ogImage` and `twitterImage` tags when the `<head>` tags are rendered. I don't want to set the `seoImage` property using the CMS or any defined fields - I need to set it conditionally as above.

**Comment by khalwat** [comment]
Yes, unless you have the `ogImage` and `twitterImage` set to pull from the `seoImage`, this is expected behavior.

What do you see if you do:

```
{{ dump(seomatic.meta.ogImage)) }}
{{ dump(seomatic.meta.twitterImage)) }}

```

**Comment by roberttolton** [comment]
Ah, yes - those I have set to inherit (selecting in the drop down menu) the SEO Image.

Regarding the dump, that’s what my previous code displayed - the values are empty. There’s nothing between the square brackets.

**Comment by khalwat** [comment]
Something isn't right here -- I'm unclear why we're using `{{ seomatic.meta.parsedValue('twitterImage') }}` -- tell you what, let's do a screen share to see if we can't get to the bottom of it?

You definitely should be able to set any of these properties to anything you like, so we can get it working. Ping me on Craft Slack to do a screen share?

**Comment by robmcfadden** [comment]
Chiming in here because I'm seeing this too. I have Twitter and Facebook set to pull from the SEO image, and they are working right if I have an SEO image set on the entry, but setting `{% do seomatic.meta.seoImage(url) %}` from the template does not work, where as setting both `{% do seomatic.meta.twitterImage(url) %}` and `{% do seomatic.meta.ogImage(url) %}` does. Not a problem for me as I'm fine setting both, though.

**Comment by khalwat** [comment]
Okay, I think this was all me being thick in not understanding what was being asked.

The following are all separate variables:
```
seomatic.meta.seoImage
seomatic.meta.twitterImage
seomatic.meta.ogImage
```

If you set them to "Same As SEO Image" in the AdminCP, it does some magic for the Twitter & Facebook URLs:

```
seomatic.meta.twitterImage -> {seomatic.helper.socialTransform(79, "twitter-large", 1, "fit")}
seomatic.meta.ogImage -> {seomatic.helper.socialTransform(79, "facebook", 1, "fit")}
```

It does this so that it can apply the appropriate transforms (if configured to do so) automagically.

But this also means that if you just change `seomatic.meta.seoImage` it is _not_ going to propagate down to `seomatic.meta.twitterImage` and `seomatic.meta.ogImage` automagically.

This is a "feature" in that it allows you to have totally different images for each social media service, while also having them appropriately transformed for each service. The downside to this is that you'll need to change _all three_ if you're doing so via Twig in a template.

While I realize that in some cases, you want to just change one variable and have that bubble down, SEOmatic can't know this.

Also @robmcfadden I'd recommend that you do what SEOmatic does, and wrap them in `seomatic.helper.socialTransform()` if the images are coming from Assets.

**Comment by khalwat** [comment]
@roberttolton I'm so sorry for not understanding what you were asking right away.

The TL;DR is that if you want to change the image via Twig in your templates, you'll need to change all three variables.

**Comment by roberttolton** [comment]
No problem, I tried showing the output of tags and code at the same time to try and help debug, but think that confused things.

Thanks for the clarification!

**Comment by khalwat** [comment]
re-opened to see if we can't make this work as expected

**Comment by LouisCuvelier** [comment]
Maybe add something in the doc about this. This issue saved me a lot of time.
